src/stash/src/postgres.rs

```diff
@@ -61,24 +64,38 @@ CREATE TABLE uppers (
 /// tables are not specified and should not be relied upon. The only promise is
 /// stability. Any changes to the table schemas will be accompanied by a clear
 pub struct Postgres {
-    conn: Client,
+    url: String,
+    client: Option<Client>,
     epoch: i64,
+    rng: [u8; 16],
 }
 
 impl std::fmt::Debug for Postgres {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.debug_struct("Postgres")
+            .field("url", &self.url)
             .field("epoch", &self.epoch)
+            .field("rng", &self.rng)
             .finish_non_exhaustive()
     }
 }
 
 impl Postgres {
     /// Opens the stash stored at the specified path.
-    pub async fn open(mut conn: Client) -> Result<Postgres, StashError> {
-        conn.batch_execute("SET default_transaction_isolation = serializable")
-            .await?;
-        let tx = conn.transaction().await?;
+    pub async fn new(url: String) -> Result<Postgres, StashError> {
+        let mut conn = Postgres {
+            url,
+            client: None,
+            epoch: 0,
+            // The call to rand::random here assumes that the rng's seed source is from a
+            // secure source that will differ per thread. The docs for ThreadRng say it "is
+            // automatically seeded from OsRng", which meets this requirement.
+            rng: rand::random(),
+        };
+        // Do the initial connection once here so we don't get stuck in transact's
+        // retry loop if the url is bad.
+        conn.connect().await?;
+        let tx = conn.client.as_mut().unwrap().transaction().await?;
```

Comment:
this `connect`-then-`unwrap` looks a little weird to me, could have have a `PostgresConfig` that has a `connect` that returns a `Postgres`?

Reply:
What's weird about it? `client` is guaranteed to be Some at this point (I've added docs saying that). Also I don't understand your suggestion. We need to be able to reconnect, or signal that we need to reconnect by setting client to None.